Subject: Replace pkg_resources with importlib.metadata in deluge and its plugins

The python3-setuptools port installs no pkg_resources module, and
deluge/ui/ui_entry.py, deluge/pluginmanagerbase.py and the common.py of
every plugin under deluge/plugins/ import it at the top, so unpatched
neither the deluge command, the plugin manager nor any plugin can be
imported.

deluge/common.py loses its pkg_resources fallback for the version lookup.
ui_entry.py reads the deluge.ui entry points through importlib.metadata.
The plugin manager walks the plugin directories itself, reads each .egg
directory, .egg-info or .dist-info it finds with Distribution.at(), and
loads the plugin from its entry points; it reads no zipped .egg file, so
a plugin installed as one into ~/.config/deluge/plugins is not found.
Each plugin's common.py takes resource_filename from deluge.common, which
has the same (package, path) arguments.

diff --git a/deluge/common.py b/deluge/common.py
index 7b76d24..4631a6a 100644
--- a/deluge/common.py
+++ b/deluge/common.py
@@ -33,10 +33,7 @@ from urllib.request import pathname2url
 from deluge.decorators import deprecated
 from deluge.error import InvalidPathError
 
-try:
-    from importlib.metadata import distribution
-except ImportError:
-    from pkg_resources import get_distribution as distribution
+from importlib.metadata import distribution
 
 
 try:
diff --git a/deluge/pluginmanagerbase.py b/deluge/pluginmanagerbase.py
index 52dae61..7798571 100644
--- a/deluge/pluginmanagerbase.py
+++ b/deluge/pluginmanagerbase.py
@@ -11,9 +11,10 @@
 
 import email
 import logging
-import os.path
+import os
+import sys
+from importlib.metadata import Distribution
 
-import pkg_resources
 from twisted.internet import defer
 from twisted.python.failure import Failure
 
@@ -102,21 +103,40 @@ class PluginManagerBase:
         ]
         plugin_dirs = [base_dir, user_dir] + base_subdir
 
-        for dirname in plugin_dirs:
-            pkg_resources.working_set.add_entry(dirname)
-        self.pkg_env = pkg_resources.Environment(
-            plugin_dirs, platform=None, python=None
-        )
-
+        self._plugin_dists = {}
         self.available_plugins = []
-        for name in self.pkg_env:
-            log.debug(
-                'Found plugin: %s %s at %s',
-                self.pkg_env[name][0].project_name,
-                self.pkg_env[name][0].version,
-                self.pkg_env[name][0].location,
-            )
-            self.available_plugins.append(self.pkg_env[name][0].project_name)
+
+        for plugin_dir in plugin_dirs:
+            if not os.path.isdir(plugin_dir):
+                continue
+            if plugin_dir not in sys.path:
+                sys.path.insert(0, plugin_dir)
+
+            for item in os.listdir(plugin_dir):
+                item_path = os.path.join(plugin_dir, item)
+                dist = None
+
+                if item.endswith('.egg') and os.path.isdir(item_path):
+                    # Add .egg directory to sys.path for imports
+                    if item_path not in sys.path:
+                        sys.path.insert(0, item_path)
+                    egg_info = os.path.join(item_path, 'EGG-INFO')
+                    if os.path.isdir(egg_info):
+                        dist = Distribution.at(egg_info)
+                elif item.endswith(('.egg-info', '.dist-info')) and os.path.isdir(item_path):
+                    dist = Distribution.at(item_path)
+
+                if dist:
+                    name = dist.metadata['Name']
+                    if name not in self._plugin_dists:
+                        self._plugin_dists[name] = dist
+                        self.available_plugins.append(name)
+                        log.debug(
+                            'Found plugin: %s %s at %s',
+                            name,
+                            dist.version,
+                            plugin_dir,
+                        )
 
     def enable_plugin(self, plugin_name):
         """Enable a plugin.
@@ -138,28 +158,27 @@ class PluginManagerBase:
             return defer.succeed(True)
 
         plugin_name = plugin_name.replace(' ', '-')
-        egg = self.pkg_env[plugin_name][0]
-        # Activate is required by non-namespace plugins.
-        egg.activate()
+        dist = self._plugin_dists[plugin_name]
         return_d = defer.succeed(True)
 
-        for name in egg.get_entry_map(self.entry_name):
+        entry_points = [ep for ep in dist.entry_points if ep.group == self.entry_name]
+        for ep in entry_points:
             try:
-                cls = egg.load_entry_point(self.entry_name, name)
+                cls = ep.load()
                 instance = cls(plugin_name.replace('-', '_'))
             except component.ComponentAlreadyRegistered as ex:
                 log.error(ex)
                 return defer.succeed(False)
             except Exception as ex:
                 log.error(
-                    'Unable to instantiate plugin %r from %r!', name, egg.location
+                    'Unable to instantiate plugin %r from %r!', ep.name, dist.metadata['Name']
                 )
                 log.exception(ex)
                 continue
             try:
                 return_d = defer.maybeDeferred(instance.enable)
             except Exception as ex:
-                log.error('Unable to enable plugin: %s', name)
+                log.error('Unable to enable plugin: %s', ep.name)
                 log.exception(ex)
                 return_d = defer.fail(False)
 
@@ -167,7 +186,7 @@ class PluginManagerBase:
                 import warnings
 
                 warnings.warn_explicit(
-                    DEPRECATION_WARNING % name,
+                    DEPRECATION_WARNING % ep.name,
                     DeprecationWarning,
                     instance.__module__,
                     0,
@@ -257,14 +276,14 @@ class PluginManagerBase:
     def get_plugin_info(self, name):
         """Returns a dictionary of plugin info from the metadata"""
 
-        if not self.pkg_env[name]:
+        if name not in self._plugin_dists:
             log.warning('Failed to retrieve info for plugin: %s', name)
             info = {}.fromkeys(METADATA_KEYS, '')
             info['Name'] = info['Version'] = 'not available'
             return info
 
-        pkg_info = self.pkg_env[name][0].get_metadata('PKG-INFO')
-        return self.parse_pkg_info(pkg_info)
+        metadata = self._plugin_dists[name].metadata
+        return {key: metadata.get(key, '') for key in METADATA_KEYS}
 
     @staticmethod
     def parse_pkg_info(pkg_info):
diff --git a/deluge/plugins/AutoAdd/deluge_autoadd/common.py b/deluge/plugins/AutoAdd/deluge_autoadd/common.py
index 6a790cb..38ba943 100644
--- a/deluge/plugins/AutoAdd/deluge_autoadd/common.py
+++ b/deluge/plugins/AutoAdd/deluge_autoadd/common.py
@@ -13,7 +13,7 @@
 
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename, subdir=False):
diff --git a/deluge/plugins/Blocklist/deluge_blocklist/common.py b/deluge/plugins/Blocklist/deluge_blocklist/common.py
index f105941..fd21eb4 100644
--- a/deluge/plugins/Blocklist/deluge_blocklist/common.py
+++ b/deluge/plugins/Blocklist/deluge_blocklist/common.py
@@ -15,7 +15,7 @@ import os.path
 from functools import wraps
 from sys import exc_info
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename):
diff --git a/deluge/plugins/Execute/deluge_execute/common.py b/deluge/plugins/Execute/deluge_execute/common.py
index eb47f13..25438dd 100644
--- a/deluge/plugins/Execute/deluge_execute/common.py
+++ b/deluge/plugins/Execute/deluge_execute/common.py
@@ -13,7 +13,7 @@
 
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename):
diff --git a/deluge/plugins/Extractor/deluge_extractor/common.py b/deluge/plugins/Extractor/deluge_extractor/common.py
index eb47f13..25438dd 100644
--- a/deluge/plugins/Extractor/deluge_extractor/common.py
+++ b/deluge/plugins/Extractor/deluge_extractor/common.py
@@ -13,7 +13,7 @@
 
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename):
diff --git a/deluge/plugins/Label/deluge_label/common.py b/deluge/plugins/Label/deluge_label/common.py
index eb47f13..25438dd 100644
--- a/deluge/plugins/Label/deluge_label/common.py
+++ b/deluge/plugins/Label/deluge_label/common.py
@@ -13,7 +13,7 @@
 
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename):
diff --git a/deluge/plugins/Notifications/deluge_notifications/common.py b/deluge/plugins/Notifications/deluge_notifications/common.py
index 9993f5c..57c1fb0 100644
--- a/deluge/plugins/Notifications/deluge_notifications/common.py
+++ b/deluge/plugins/Notifications/deluge_notifications/common.py
@@ -14,7 +14,7 @@
 import logging
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 from twisted.internet import defer
 
 from deluge import component
diff --git a/deluge/plugins/Scheduler/deluge_scheduler/common.py b/deluge/plugins/Scheduler/deluge_scheduler/common.py
index eb47f13..25438dd 100644
--- a/deluge/plugins/Scheduler/deluge_scheduler/common.py
+++ b/deluge/plugins/Scheduler/deluge_scheduler/common.py
@@ -13,7 +13,7 @@
 
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename):
diff --git a/deluge/plugins/Stats/deluge_stats/common.py b/deluge/plugins/Stats/deluge_stats/common.py
index eb47f13..25438dd 100644
--- a/deluge/plugins/Stats/deluge_stats/common.py
+++ b/deluge/plugins/Stats/deluge_stats/common.py
@@ -13,7 +13,7 @@
 
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename):
diff --git a/deluge/plugins/Toggle/deluge_toggle/common.py b/deluge/plugins/Toggle/deluge_toggle/common.py
index eb47f13..25438dd 100644
--- a/deluge/plugins/Toggle/deluge_toggle/common.py
+++ b/deluge/plugins/Toggle/deluge_toggle/common.py
@@ -13,7 +13,7 @@
 
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename):
diff --git a/deluge/plugins/WebUi/deluge_webui/common.py b/deluge/plugins/WebUi/deluge_webui/common.py
index eb47f13..25438dd 100644
--- a/deluge/plugins/WebUi/deluge_webui/common.py
+++ b/deluge/plugins/WebUi/deluge_webui/common.py
@@ -13,7 +13,7 @@
 
 import os.path
 
-from pkg_resources import resource_filename
+from deluge.common import resource_filename
 
 
 def get_resource(filename):
diff --git a/deluge/ui/ui_entry.py b/deluge/ui/ui_entry.py
index f0a9831..2cf7e4b 100644
--- a/deluge/ui/ui_entry.py
+++ b/deluge/ui/ui_entry.py
@@ -16,8 +16,7 @@ import argparse
 import logging
 import os
 import sys
-
-import pkg_resources
+from importlib.metadata import entry_points
 
 import deluge.common
 import deluge.configmanager
@@ -35,7 +34,7 @@ def start_ui():
 
     # Get the registered UI entry points
     ui_entrypoints = {}
-    for entrypoint in pkg_resources.iter_entry_points('deluge.ui'):
+    for entrypoint in entry_points(group='deluge.ui'):
         try:
             ui_entrypoints[entrypoint.name] = entrypoint.load()
         except ImportError:
